### PR TITLE
Jira Rest Worklog User Helper

### PR DESCRIPTION
**Pull request by spheresh**

### CODE REVIEW

File: drupal/modules/custom/intranet_jira/intranet_jira.module [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+use Drupal\intranet_jira\IntranetJiraApiRest;
+use Drupal\intranet_jira\IntranetJiraProjectTask;
+
+/**
+ * Add jira field to user entity
+ *
+ * @param $fields
+ * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+ */
+function intranet_jira_entity_base_field_info_alter(&$fields, EntityTypeInterface $entity_type) {
```
Comment (PavelLoparev) [inline drupal/modules/custom/intranet_jira/intranet_jira.module]: Maybe we need real field in user account?

File: drupal/modules/custom/intranet_jira/intranet_jira.module [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    //->condition('field_jira_started', date("Y-m-d", REQUEST_TIME), '<=');
+  $nids = $query->execute();
+  if ($nids) {
+    $nodes = entity_load_multiple('node', $nids);
+    foreach ($nodes as $node) {
+      $node->delete();
+    }
+  }
+
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/intranet_jira.module]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  /**
+   * @var array
+   */
+  private $conditional;
+
+  /**
+   * IntranetJiraApiSearch constructor.
+   */
+  public function __construct() {
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php]: empty constructor not needed

File: drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  /**
+   * Calculate condition by relative day (ex. +1 day)
+   *
+   * @param $day
+   * @return \DateTime
+   */
+  public function calcDay($day) {
+    $action = "date" . (($day >= 0) ? "_add" : "_sub");
+    $day = abs($day);
+    $interval = (string)$day . " days";
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php]: space needed after cast expression

File: drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  /**
+   * Convert to string
+   *
+   * @return string
+   */
+  public function __toString() {
+    return implode(" AND ", $this->conditional);
+  }
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraApiSearch.php]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    return TRUE;
+  }
+
+  public function refreshItems() {
+    $jira = \Drupal::service("intranet_jira.api_rest");
+    $worklog = $jira->getWorklog($this);
+    $worklog->save();
+  }
+
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  public function timeExists() {
+    $time = $this->storage->getProjectTask($this);
+    /**
+     * @TODO Rename check function
+     */
+    return ($time->isNonExists() ? FALSE : TRUE);
+  }
+
+  public function getUpdated() {
+    return strtotime(substr($this->task->fields->updated, 0 ,19));
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php]: TODO: check this (maybe strtotime($this->task->fields->updated) ?)

File: drupal/modules/custom/intranet_helper/src/Controller/IntranetUserController.php [hunk 8]
```diff
@@ -137,7 +137,14 @@ public function checkState(Request $request) {
       $jira = (boolean) $account->field_jira_required->value;
     }
     else {
-      // TODO: get '$jira' value from jira.
+      // @TODO TEST MODE
+      $need_update = TRUE;
+      $jira_rest = \Drupal::service("intranet_jira.api_rest");
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_helper/src/Controller/IntranetUserController.php]: I think $jira_rest should be protected variable in controller class and should be initialized in constructor

File: drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    $time = $this->storage->getProjectTask($this);
+
+    if(!$time->isNonExists() && $this->getUpdated() != $time->getValue()) {
+      return FALSE;
+    }
+    return TRUE;
+  }
+
+  public function refreshItems() {
+    $jira = \Drupal::service("intranet_jira.api_rest");
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraProjectTask.php]: I think $jira_rest should be protected variable in controller class and should be initialized in constructor

File: drupal/modules/custom/intranet_jira/src/IntranetJiraResponse.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * (PHP 5 &gt;= 5.0.0)<br/>
+   * Rewind the Iterator to the first element
+   * @link http://php.net/manual/en/iterator.rewind.php
+   * @return void Any returned value is ignored.
+   */
+  public function rewind() {
+    reset($this->response->issues);
+  }
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraResponse.php]: EOF

File: drupal/modules/custom/intranet_jira/src/Controller/IntranetJiraTestRouter.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+use Drupal\intranet_jira\Plugin\QueueWorker\IntranetJiraAggregator;
+
+/**
+ * Class IntranetJiraTestRouter.
+ *
+ * @package Drupal\intranet_jira\Controller
+ */
+class IntranetJiraTestRouter extends ControllerBase {
+
+  const SETTINGS_NAME = 'modulename.settings';
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/Controller/IntranetJiraTestRouter.php]: change config file name

File: drupal/modules/custom/intranet_jira/src/IntranetJiraStorage.php [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+ * @package Drupal\intranet_jira
+ */
+class IntranetJiraStorage {
+
+  /**
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  private $storage;
+
+  const SETTINGS_NAME = 'modulename.settings';
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraStorage.php]: change config file name

File: drupal/modules/custom/intranet_jira/src/IntranetJiraStorage.php [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+      ->condition('field_jira_started', $format);
+    $nids = $query->execute();
+    return (count($nids) > 0 ? TRUE : FALSE);
+  }
+  /**
+   * @param IntranetJiraWorklog $worklog_class
+   */
+  public function storeWorkLog($worklog_class) {
+    // @TODO
+    if(time() - (86400 * 4) > strtotime($worklog_class->started())) {
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraStorage.php]: 86400 value should be const in class definition for example ONE_DAY

File: drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  /**
+   * Retrun the value of time
+   * @return string
+   */
+  public function getValue() {
+    return $this->time->value;
+  }
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+   */
+  public function __construct($time) {
+    $this->time = $time;
+  }
+
+  /**
+   * Check that the value is empty
+   * @return bool
+   */
+  public function isNonExists() {
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php]: isNonExists -> isNotExists

File: drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    return is_null($this->time->value);
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraTime.php]: return empty($this->time->value);

File: drupal/modules/custom/intranet_jira/src/IntranetJiraWorkLogResponse.php [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  public function save() {
+    foreach ($this->response->worklogs as $worklog) {
+      $worklog_class = new IntranetJiraWorklog($worklog, $this->task);
+
+      $this->storage->storeWorkLog($worklog_class);
+      //IntranetJiraStorageTask::storeWorkLog($worklog_class);
+    }
+
+  }
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraWorkLogResponse.php]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraWorklog.php [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  }
+  public function author() {
+    return $this->worklog->author->name;
+  }
+  public function started() {
+    return substr($this->worklog->started, 0, 10);
+  }
+
+
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraWorklog.php]: EOF

File: drupal/modules/custom/intranet_jira/src/IntranetJiraWorklog.php [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    return $this->worklog->id;
+  }
+  public function timeSpent() {
+    return $this->worklog->timeSpentSeconds;
+  }
+  public function author() {
+    return $this->worklog->author->name;
+  }
+  public function started() {
+    return substr($this->worklog->started, 0, 10);
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/IntranetJiraWorklog.php]: what if format will change?

File: drupal/modules/custom/intranet_jira/src/Plugin/QueueWorker/IntranetJiraAggregator.php [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+      /**
+       * @var IntranetJiraProjectTask $data
+       */
+      $project_task->setTime();
+      $project_task->refreshItems();
+    }
+  }
+
+}
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/src/Plugin/QueueWorker/IntranetJiraAggregator.php]: EOF

File: drupal/reinstall.sh [hunk 21]
```diff
@@ -5,4 +5,4 @@
 # sudo apt-get update
 # sudo apt-get install ansible
 # sudo apt-get install python-mysqldb
-time ansible-playbook -vvvv reinstall.yml --tags=all --extra-vars="installation_profile_name=pp mysql_db=drupal2"
+time ansible-playbook -vvvv reinstall.yml --tags=all --extra-vars="installation_profile_name=pp mysql_db=drupal4"
```
Comment (Propeople-ua) [inline drupal/reinstall.sh]: checkout this! it needs to be mysql_db=drupal

File: drupal/modules/custom/intranet_jira/intranet_jira.module [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+  /**
+   * Clean up old worklogs
+   */
+  $query = \Drupal::entityQuery('node')
+    ->condition('type', 'jira_worklog')
+    ->condition('field_jira_started', date("Y-m-d", REQUEST_TIME - (86400 * 4)), '<');
+    //->condition('field_jira_started', date("Y-m-d", REQUEST_TIME), '<=');
+  $nids = $query->execute();
+  if ($nids) {
+    $nodes = entity_load_multiple('node', $nids);
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/intranet_jira.module]: entity_load_multiple is deprecated in D8. use \Drupal::entityManager()->getStorage($entity_type)->loadMultiple($id) instead

File: drupal/modules/custom/intranet_jira/intranet_jira.routing.yml [hunk 23]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/intranet_jira.routing.yml]: this line not needed

File: drupal/modules/custom/intranet_jira/intranet_jira.services.yml [hunk 24]
```diff
@@ -0,0 +1,9 @@
+services:
+  intranet_jira.api_rest:
+    class: Drupal\intranet_jira\IntranetJiraApiRest
+    arguments: ["@jira_rest.jira_rest_service"]
+
+  intranet_jira.storage:
+    class: Drupal\intranet_jira\IntranetJiraStorage
+    arguments: []
+
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/intranet_jira.services.yml]: this line not needed

File: drupal/modules/custom/intranet_jira/intranet_jira.module [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+      $queue->createItem($project_task);
+    }
+  }
+
+  /**
+   * Clean up old worklogs
+   */
+  $query = \Drupal::entityQuery('node')
+    ->condition('type', 'jira_worklog')
+    ->condition('field_jira_started', date("Y-m-d", REQUEST_TIME - (86400 * 4)), '<');
```
Comment (Propeople-ua) [inline drupal/modules/custom/intranet_jira/intranet_jira.module]: 86400 value should be const in class definition. For example ONE_DAY

**Pull request merged**